currency_monthly_rate/models/res_currency.py

```diff
@@ -0,0 +1,95 @@
+# Copyright 2018 Camptocamp SA
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+from odoo import models, fields, api, _
+
+
+class ResCurrency(models.Model):
+    _inherit = "res.currency"
+
+    monthly_rate = fields.Float(compute='_compute_current_monthly_rate',
+                                string='Current Monthly Rate', digits=(12, 6),
+                                help='The monthly rate of the currency to '
+                                     'the currency of rate 1.')
+    monthly_rate_ids = fields.One2many('res.currency.rate.monthly',
+                                       'currency_id', string='Monthly rates')
+
+    @api.multi
+    def _compute_current_monthly_rate(self):
+        date = self.env.context.get('date') or fields.Date.today()
+        company_id = self.env.context.get('company_id') or self.env[
+            'res.users']._get_company().id
+        # the subquery selects the first rate before 'date' for the given
+        # currency/company
+        query = """SELECT c.id, (SELECT r.rate
+                                 FROM res_currency_rate_monthly r
+                                 WHERE r.currency_id = c.id
+                                 AND r.name <= %s
+                                 AND (r.company_id IS NULL
+                                      OR r.company_id = %s)
+                                 ORDER BY r.company_id, r.name DESC
+                                 LIMIT 1) AS rate
+                   FROM res_currency c
+                   WHERE c.id IN %s"""
+        self._cr.execute(query, (date, company_id, tuple(self.ids)))
+        currency_rates = dict(self._cr.fetchall())
+        for currency in self:
+            currency.monthly_rate = currency_rates.get(currency.id) or 1.0
+
+    @api.model
+    def _get_conversion_rate(self, from_currency, to_currency):
+        monthly = self.env.context.get('monthly_rate')
+        if not monthly:
+            return super()._get_conversion_rate(from_currency, to_currency)
+        from_currency = from_currency.with_env(self.env)
+        to_currency = to_currency.with_env(self.env)
+        return to_currency.monthly_rate / from_currency.monthly_rate
```

Comment:
Are we certain there will be no `div by 0`?

Reply:
Odoo allows to set 0 as `rate` for `res.currency.rate`. However as `rate` is computed on `res.currency`, a 0 value on the actual `res.currency.rate` will be computed as 1 on `res.currency` : https://github.com/odoo/odoo/blob/11.0/odoo/addons/base/res/res_currency.py#L61
Therefore we used the same behaviour for monthly rates.